packages/wdio-firefox-profile-service/src/launcher.js

```diff
@@ -0,0 +1,83 @@
+import Profile from 'firefox-profile'
+import { promisify } from 'util'
+
+export default class FirefoxProfileLauncher {
+    async onPrepare (config, capabilities) {
+        this.config = config
+        this.capabilities = capabilities
+
+        // Return if no profile options were specified
+        if (!this.config.firefoxProfile) {
+            return
+        }
+
+        this.profile = new Profile()
+
+        // Set preferences and proxy
+        this._setPreferences()
+
+        if (!Array.isArray(this.config.firefoxProfile.extensions)) {
+            return this._buildExtension()
+        }
+
+        // Add the extension
+        await promisify(::this.profile.addExtensions)(this.config.firefoxProfile.extensions)
+
+        return this._buildExtension()
+    }
+
+    /**
+     * Sets any preferences and proxy
+     */
+    _setPreferences () {
+        for (const preference in this.config.firefoxProfile) {
+            if (['extensions', 'proxy'].includes(preference)) {
+                continue
+            }
+
+            this.profile.setPreference(preference, this.config.firefoxProfile[preference])
+        }
+
+        if (this.config.firefoxProfile.proxy) {
+            this.profile.setProxy(this.config.firefoxProfile.proxy)
+        }
+
+        this.profile.updatePreferences()
+    }
+
+    async _buildExtension () {
+        const zippedProfile = await promisify(::this.profile.encoded)()
+
+        if (Array.isArray(this.capabilities)) {
+            this.capabilities
+                .filter((capability) => capability.browserName === 'firefox')
+                .forEach((capability) => {
+                    // for older firefox and geckodriver versions
+                    capability.firefox_profile = zippedProfile
+
+                    // for firefox >= 56.0 and geckodriver >= 0.19.0
+                    capability['moz:firefoxOptions'] = {
+                        profile: zippedProfile,
+                    }
+                })
+
+            return
+        }
+
+        for (const browser in this.capabilities) {
+            const capability = this.capabilities[browser].desiredCapabilities
+
+            if (!capability || capability.browserName !== 'firefox') {
+                continue
+            }
+
+            // for older firefox and geckodriver versions
```

Comment:
similar as do what in line 55 happens .. I doubt this logic will change much so let's move the assignments into a separate function